### ISSUE TITLE
Quick preview

### ISSUE THREAD

**Issue by paweljarosz** [defold/defold#8373]
_Is it a private platform specific feature? (i.e. would it break your NDA?)<br>
Then please add your feature request in the corresponding extension repository!_

**Is your feature request related to a problem? Please describe (REQUIRED):**

**Describe the solution you'd like (REQUIRED):**

**Describe alternatives you've considered (REQUIRED):**

**Additional context (OPTIONAL):**

**Comment by britzl** [comment]
I like this idea. One question to figure out is how to play the sound? Should the editor bundle standalone exes to play the sounds or is there some kind of Java lib to play ogg and wav files?

**Comment by AGulev** [comment]
> I like this idea. One question to figure out is how to play the sound? Should the editor bundle standalone exes to play the sounds or is there some kind of Java lib to play ogg and wav files?

Maybe we can use javafx mediaplayer:
```
import javafx.scene.media.Media;
import javafx.scene.media.MediaPlayer;
```
https://jenkov.com/tutorials/javafx/media.html

**Comment by AGulev** [comment]
> import javafx.scene.media.Media;
> import javafx.scene.media.MediaPlayer;

it doesn't support `ogg` =/

**Comment by wavesinaroom** [comment]
Hi @britzl! 

I'd like to start working on this issue to get familiar with the project. I've got an idea for this feature

Obviously we need a button to interact with but we could use the same button to show notifications meaning:

You click on the button and "Playing" for a clip assigned to the component, "Empty" in the absence of the clip and "Audio clipping". Why those three messages? First one is self-explanatory but the second and the third help dev to notice what the issue is without running the game. It's really easy and common to try to play a component without an audio file an get a reference exception or clipping audio even for trained ears. Now the second time you click on the button you stop playback, it's simple yet informative.

How about calling this feature audio preview? Not entirely accurate but it's a word that could be used for other contexts than visuals as far as I know.

@AGulev I had a quick look at the editor folder but I got scared with seeing clojure files ;). Never touched clojure in my whole life so what would you suggest me to start learning it with to be dangerous coding with it?

What workflow do you use guys for designing a piece of UI for the editor? Do we have a designer? Should we handle this by ourselves?

**Comment by britzl** [comment]
> Obviously we need a button to interact with

Probably, but not necessarily.

* If you select a sprite in the Outline and press Space you preview the currently selected animation. You could do the same when a sound component is selected.
* There could also be a Play/Preview button in the Properties panel. 
* But perhaps it is also important to be able to preview sound files before assigning them to a sound component? Ie from the Asset panel? In that case selecting a .wav, .ogg or .opus file and hitting space could preview it?

> What workflow do you use guys for designing a piece of UI for the editor? Do we have a designer? Should we handle this by ourselves?

We do not have a designer but try to follow common design principles in other software and existing design principles in the Defold editor.

> One question to figure out is how to play the sound?

I think this is an important question to answer. How are you going to play the sound? I'd like to see some solutions to this first.

**Comment by wavesinaroom** [comment]
>One question to figure out is how to play the sound? 

Good news! Apparently there's a Java package called [javax.sound.sampled](https://docs.oracle.com/javase/tutorial/sound/sampled-overview.html) we can import to play sounds through its `Clip` object. `Clip` has `Start`, `Stop` and `Pause` methods which are enough for us to make the audio tester, a name inspired by @paweljarosz bug description. I'll take some time looking into this package to understand how it works.

>If you select a sprite in the Outline and press Space you preview the currently selected animation. You could do the same when a sound component is selected.

I'm positive to find a solution with `javax.sound.sampled` package. Although this solution would be restricted for assets as I see it. Why? Well, here we go with the problem. We want to create randomizers for pitch and volume as well as a clip randomizer and a clip sequencer (if I remember correctly). Testing those actions should go through the `AudioManager` we create through `sound.cpp` and friends (Web audio API, Android support). So, if we want to test audio that goes through this processes we have to go in another direction possibly. 

Let's keep this simple for now I'd like to suggest by starting with trying to test audio by pressing on the speaker icon we've got next to the sound component. We could change its color to indicate the state (Playing -> green, Empty, clipping -> red). Of course this is just an Idea that I need to see if it's possible with `fxml` and `css`. 

We could also add a tooltip for giving usage instruction but again I need to see if that's possible with `fxml` and `css`.

**Comment by britzl** [comment]
> Good news! Apparently there's a Java package called [javax.sound.sampled](https://docs.oracle.com/javase/tutorial/sound/sampled-overview.html) we can import to play sounds through its `Clip` object

Does it support the formats we need? wav, ogg, opus?

**Comment by wavesinaroom** [comment]
Good question, I tested the formats we need with the following code:

```java
import java.io.File;
import java.io.IOException;
import java.net.URL;
import javax.sound.sampled.*;

public class FormatTester {

    public static void main(String[] args)
        throws UnsupportedAudioFileException, IOException, LineUnavailableException {
        String[] paths = { "music.ogg", "music.opus", "door_opening.wav" };

        for (String path : paths) {
            try {
                AudioInputStream audioInputStream =
                    AudioSystem.getAudioInputStream(
                        new File(path).getAbsoluteFile()
                    );
                Clip clip = AudioSystem.getClip();
                clip.open(audioInputStream);
            } catch (UnsupportedAudioFileException e) {
                e.printStackTrace();
            }
        }
    }
}

```

Unfortunately, I got `UnsupportedAudioFileExection` for ogg, opus

```
Testing Audio format
Testing: music.ogg
javax.sound.sampled.UnsupportedAudioFileException: File of unsupported format
	at java.desktop/javax.sound.sampled.AudioSystem.getAudioInputStream(AudioSystem.java:1046)
	at FormatTester.main(FormatTester.java:17)
Testing: music.opus
javax.sound.sampled.UnsupportedAudioFileException: File of unsupported format
	at java.desktop/javax.sound.sampled.AudioSystem.getAudioInputStream(AudioSystem.java:1046)
	at FormatTester.main(FormatTester.java:17)
```

wav format was supported but threw another exception because this is my first time writing Java. I'll try to find another way to be able to playback audio that supports the formats we need. Ogg and Opus aren't there.

Supported formats are listed in the [Java Sound Technology documentation](https://docs.oracle.com/javase/8/docs/technotes/guides/sound/index.html)

**Comment by britzl** [comment]
> Unfortunately, I got `UnsupportedAudioFileExection` for ogg, opus

It's probably possible to find implementations for ogg and opus which can hook into javax.sound.sampled.

- https://mvnrepository.com/artifact/com.googlecode.soundlibs/vorbisspi
- https://github.com/ymnk/jorbis

One other alternative might be to call out to some native OS function to play the sound. Another might be to package the Defold sound library as a stand alone library and call into it from Java using JNI.

**Comment by wavesinaroom** [comment]
Thanks for your ideas Björn! I'm checking your links now. However, I find packaging the Defold sound library a good option because it would keep consistency between playback testing and the actual audio implementation in the game. I was concerned about having the same code twice in two languages. Let me check the native OS function as well, I think it's possible to play files but if we want to audition more advanced implementation such as pitch randomization, we may no be able to rely on the OS function. These are assumption I want to confirm now.

**Comment by wavesinaroom** [comment]
Alright, I think I'm inclined to:

>Another might be to package the Defold sound library as a stand alone library and call into it from Java using JNI.

If we can run Defold engine audio system on editor startup and use its methods using JNI, we would be able to call the same functions the engine calls meaning that testing audio would be done through the same system the game uses. So, you're basically right. 

I read [this article](https://medium.com/@Neelesh-Janga/run-native-code-in-a-java-environment-using-jni-0dac9c3e6c39) that explains how to call C/C++ in Java environment. It gives a clear and simple implementation but I'd like to check how it could be apply to our source code.

**Comment by wavesinaroom** [comment]
I asked some technical questions on Discord after spending some time reading the source code. While we get answers to them, I think we can discuss about the design of the feature. 

I agree we should have an option to play the file from the assets panel be it a shortcut or maybe with a context menu. However, I think we can also have an option to play how we can play sound and its changes from the sound component. Why? Well, in the future we might need to audition how the sound pitch and volume is randomized. Besides, there are containers listed in the sound milestone that could also be audition by clicking on the speaker icon.

In summary, we can audition sound on the editor with two actions:

1. Asset panel: Plays the sound file only as a way to let the user identify which asset he/she wants to use in a sound component.

2. Sound component. Plays the sound, its changes (pitch and volume) plus containers. For now though, we'll just hear the sound file played like in the asset panel

**Comment by mathias-a-la-basse** [comment]
Hi, personally I would like a small "play icon" beside the asset or sound component named to just click and listen the sound. with a play/pause icon change to see that the sound was or is playing
